### ISSUE TITLE
Testing on training data for era 1 never stops

### ISSUE THREAD

**Issue by Zhen-hao** [zhangxiangxiao/Crepe#3]
It seems that the part "Testing on training data for era 1 never stops" doesn't stop on the example data. 
The function Test:run(logfunc) starts a loop. I have no idea how to control this loop. 
What parameters should change in config.lua? 

Thanks in advance!

**Comment by zhangxiangxiao** [comment]
The loop should be ended when testing is done in all training data. It may take a long time since there are many training samples.

**Comment by Zhen-hao** [comment]
Thanks for your reply! The thing is that I cut the training data to only 20000 rows, yet the Testing on training data doesn't seem to stop. Below is a copy from I got:

Training for era 1
epo: 1, rat: 1.00e-02, err: 0.00e+00, obj: 2.64e+00, dat: 3.97e-01, fpp: 2.73e+00, bpp: 3.90e+00, upd: 1.38e-02
epo: 3, rat: 1.00e-02, err: 0.00e+00, obj: 2.61e+00, dat: 2.22e-01, fpp: 2.29e+00, bpp: 3.90e+00, upd: 1.35e-02
epo: 5, rat: 1.00e-02, err: 0.00e+00, obj: 2.56e+00, dat: 2.19e-01, fpp: 2.35e+00, bpp: 3.93e+00, upd: 1.37e-02
epo: 7, rat: 1.00e-02, err: 0.00e+00, obj: 2.48e+00, dat: 2.36e-01, fpp: 2.34e+00, bpp: 3.91e+00, upd: 1.35e-02
epo: 9, rat: 1.00e-02, err: 0.00e+00, obj: 2.38e+00, dat: 3.43e-01, fpp: 2.81e+00, bpp: 4.19e+00, upd: 1.36e-02
Disabling dropouts
Testing on training data for era 1
n: 128, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 6.32e-01, fpp: 2.78e+00, acc: 8.19e-04
n: 512, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.50e-01, fpp: 2.75e+00, acc: 7.24e-04
n: 1024, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.48e-01, fpp: 2.76e+00, acc: 7.26e-04
n: 1536, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 2.50e-01, fpp: 2.46e+00, acc: 4.14e-04
n: 2048, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 4.89e-01, fpp: 2.30e+00, acc: 3.52e-04
n: 2560, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 1.99e-01, fpp: 2.30e+00, acc: 3.72e-04
n: 3072, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 2.20e-01, fpp: 2.33e+00, acc: 3.66e-04
n: 3584, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 2.08e-01, fpp: 2.32e+00, acc: 3.51e-04
n: 4096, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 2.43e-01, fpp: 2.32e+00, acc: 3.56e-04
n: 4608, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 2.29e-01, fpp: 2.32e+00, acc: 4.28e-04
n: 5120, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.09e-01, fpp: 2.73e+00, acc: 7.43e-04
n: 5632, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.49e-01, fpp: 2.78e+00, acc: 7.25e-04
n: 6016, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.56e-01, fpp: 2.84e+00, acc: 7.37e-04
Collecting garbage at n = 6400
n: 6400, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.47e-01, fpp: 2.63e+00, acc: 7.44e-04
n: 6912, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.07e-01, fpp: 2.83e+00, acc: 6.93e-04
n: 7424, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.09e-01, fpp: 2.78e+00, acc: 7.06e-04
n: 7808, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.13e-01, fpp: 2.86e+00, acc: 8.07e-04
n: 8320, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.10e-01, fpp: 2.80e+00, acc: 7.04e-04
n: 8832, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.44e-01, fpp: 2.79e+00, acc: 7.17e-04
n: 9344, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.41e-01, fpp: 2.81e+00, acc: 7.18e-04
n: 9728, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.08e-01, fpp: 2.75e+00, acc: 7.35e-04
n: 10240, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.09e-01, fpp: 2.80e+00, acc: 9.80e-03
n: 10752, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.07e-01, fpp: 2.81e+00, acc: 7.01e-04
n: 11264, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.10e-01, fpp: 2.77e+00, acc: 7.05e-04
n: 11776, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.42e-01, fpp: 2.73e+00, acc: 7.09e-04
n: 12160, e: 0.00e+00, l: 2.26e+00, err: 0.00e+00, obj: 2.26e+00, dat: 3.46e-01, fpp: 2.80e+00, acc: 7.18e-04

**Comment by felixsmueller** [comment]
Hi Zhen

what hardware do you use? It looks to me that in particular the forward and backward prop durations are quite lare.
I use the following hardware and software: 
Rackmount Server: 2 x 10 Core Intel Xeon E5-2660v3 2.6GHz Processor 64GB DDR4 2133MHz ECC REG Server Memory
2 x nVidia TESLA K80 24GB GPU Computing Processor (two GPUs per card)
Ubuntu 14.04 LTS x86_64
NVIDIA driver 355.11
CUDA 7.5.18
cuDNN 3.0
Torch branch master, commit ff7e06ca2ac7ad2d1871eae57a327d9943a4b0d7
Crepe branch master, commit 9c2dfcdf8bd19ffe2d07ae3122085177e9609645 

I get the following durations:
epo: 1, rat: 1.00e-02, err: 9.53e-01, obj: 4.96e+00, dat: 3.33e-01, fpp: 1.40e+00, bpp: 1.45e+00, upd: 3.56e-03
epo: 3, rat: 1.00e-02, err: 9.22e-01, obj: 3.20e+00, dat: 1.40e-01, fpp: 1.53e+00, bpp: 1.46e+00, upd: 3.66e-03
epo: 5, rat: 1.00e-02, err: 8.98e-01, obj: 2.77e+00, dat: 1.42e-01, fpp: 1.53e+00, bpp: 1.45e+00, upd: 3.67e-03
epo: 7, rat: 1.00e-02, err: 8.98e-01, obj: 2.88e+00, dat: 1.67e-01, fpp: 1.53e+00, bpp: 1.45e+00, upd: 3.47e-03
epo: 9, rat: 1.00e-02, err: 9.45e-01, obj: 2.76e+00, dat: 1.36e-01, fpp: 1.53e+00, bpp: 1.45e+00, upd: 3.47e-03
...
epo: 9995, rat: 1.00e-02, err: 1.56e-02, obj: 5.95e-02, dat: 1.71e-01, fpp: 1.52e+00, bpp: 1.42e+00, upd: 3.65e-03
epo: 9997, rat: 1.00e-02, err: 3.12e-02, obj: 5.85e-02, dat: 1.59e-01, fpp: 1.53e+00, bpp: 1.42e+00, upd: 3.69e-03
epo: 9999, rat: 1.00e-02, err: 4.69e-02, obj: 1.41e-01, dat: 1.45e-01, fpp: 1.52e+00, bpp: 1.42e+00, upd: 3.72e-03
Collecting garbage at epoch = 10000
Disabling dropouts
Testing on training data for era 2
n: 128, e: 7.81e-02, l: 3.00e-01, err: 7.81e-02, obj: 3.00e-01, dat: 1.54e-01, fpp: 1.52e+00, acc: 2.66e-04
n: 512, e: 5.66e-02, l: 2.10e-01, err: 5.47e-02, obj: 1.98e-01, dat: 1.53e-01, fpp: 1.52e+00, acc: 2.95e-04
n: 896, e: 5.25e-02, l: 2.09e-01, err: 7.81e-02, obj: 2.54e-01, dat: 1.63e-01, fpp: 1.52e+00, acc: 3.13e-04
...
n: 69120, e: 3.26e-02, l: 1.19e-01, err: 2.34e-02, obj: 1.25e-01, dat: 1.59e-01, fpp: 1.53e+00, acc: 3.06e-04
n: 69504, e: 3.26e-02, l: 1.19e-01, err: 2.34e-02, obj: 7.46e-02, dat: 1.43e-01, fpp: 1.53e+00, acc: 3.28e-04
n: 69888, e: 3.28e-02, l: 1.20e-01, err: 5.47e-02, obj: 1.91e-01, dat: 1.63e-01, fpp: 1.53e+00, acc: 3.03e-04

Note that the obj-duration in training gets considerably smaller over time, so that fpp and bpp dominate. Not sure why it's so big initially.

<b>Can others post as well their runtimes? </b>
According to Xiang it should only take about 2-3h per era on the small model but full dataset of DBPedia (I assume this includes training and testing). On my K80 it takes about 8h.

Regards, Felix

**Comment by zhangxiangxiao** [comment]
obj is the loss at current minibatch, not time.

**Comment by zhangxiangxiao** [comment]
Closing because of no follow-up.
